### PR TITLE
(feat) O3-3367: Add support for person attributes

### PR DESCRIPTION
**Pull request by gitcliff**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary
This PR is adding support for capturing person attributes 
Here is a sample form schema for capturing person attributes of string attributeType 
```
{
  "name": "New test form ",
  "pages": [
    {
      "label": "newPage ",
      "sections": [
        {
          "label": "New section",
          "isExpanded": "true",
          "questions": [
            {
              "label": "Person attribute (String)",
              "type": "personAttribute",
              "required": false,
              "id": "bloodTyping",
              "questionOptions": {
                "rendering": "text",
                "attributeType": "8d871d18-c2cc-11de-8d13-0010c6dffd0f"
              }
            }
          ]
        }
      ]
    }
  ],
  "processor": "EncounterFormProcessor",
  "encounterType": "dd528487-82a5-4082-9c72-ed246bd49591",
  "referencedForms": [],
  "uuid": "c0112926-506d-4a38-9ae3-10e7f9987028",
  "description": "Form that keeps HIV patient data",
  "version": "1.0"
}
```
Here is a sample form schema for capturing person attributes with attributeType of coded values

```
{
  "name": "New test form 2 ",
  "pages": [
    {
      "label": "newPage2 ",
      "sections": [
        {
          "label": "New section",
          "isExpanded": "true",
          "questions": [
            {
              "label": "Person attribute (location drop down)",
              "type": "personAttribute",
              "required": false,
              "id": "bloodTyping",
              "questionOptions": {
                "rendering": "ui-select-extended",
                "attributeType": "8d87236c-c2cc-11de-8d13-0010c6dffd0f",
                "datasource": {
                  "name": "person_attribute_location_datasource"
                }
              },
              "validators": []
            }
          ]
        }
      ]
    }
  ],
  "processor": "EncounterFormProcessor",
  "encounterType": "dd528487-82a5-4082-9c72-ed246bd49591",
  "referencedForms": [],
  "uuid": "e3452e71-16a7-42f1-bb5f-161492c153f8",
  "description": "Form that keeps HIV patient data",
  "version": "1.0"
}
```
## Screenshots
https://www.loom.com/share/c3cfa5abd10b4526aabd4017a9bee0fc

## Related Issue
https://openmrs.atlassian.net/browse/O3-3367

### CODE REVIEW

**Comment by samuelmale** [comment]
@gitcliff are still working on this?

**Comment by gitcliff** [comment]
> @gitcliff are still working on this?

yes @samuelmale..
thanks

File: src/transformers/default-schema-transformer.ts [hunk 1]
```diff
@@ -129,6 +130,29 @@ function transformByType(question: FormField) {
         ? 'date'
         : question.questionOptions.rendering;
       break;
+    case 'personAttribute':
+      handlePersonAttributeType(question);
+      break;
+  }
+}
+
+async function handlePersonAttributeType(question: FormField) {
```
Comment (samuelmale) [inline src/transformers/default-schema-transformer.ts:139]: The asynchronous nature of this function means that the encompassing function (`transform(form: FormSchema)`) should become async.

File: src/transformers/default-schema-transformer.ts [hunk 2]
```diff
@@ -129,6 +130,29 @@ function transformByType(question: FormField) {
...
+  if (question.questionOptions.rendering !== 'text') {
+    question.questionOptions.rendering === 'ui-select-extended';
+  }
+  const inputFormat = await getPersonAttributeTypeFormat(question.questionOptions?.attributeType);
+  const obj = inputFormat;
+  if (obj?.format === 'org.openmrs.Location') {
```
Comment (samuelmale) [inline src/transformers/default-schema-transformer.ts:145]:
```suggestion
  const attributeTypeFormat = await getPersonAttributeTypeFormat(question.questionOptions?.attributeType);
  if (attributeTypeFormat?.format === 'org.openmrs.Location') {
```

File: src/transformers/default-schema-transformer.ts [hunk 3]
```diff
@@ -129,6 +130,29 @@ function transformByType(question: FormField) {
...
+    question.questionOptions.datasource = {
+      name: 'person_attribute_location_datasource',
+    };
+  } else {
```
Comment (samuelmale) [inline src/transformers/default-schema-transformer.ts:149]: What happens if the configured format is not a concept? Maybe a "org.openmrs.util.AttributableDate"?

**Review by samuelmale — CHANGES_REQUESTED** [review]
I have tested this out but the form is not correctly repopulated in "edit" for "ui-select-extended" rendering.

**Review by CynthiaKamau — COMMENTED** [review]
Thanks @gitcliff for this, is it possible to add an example of how the sample schema will look ?

File: src/transformers/default-schema-transformer.ts (hunk 3 above)
Comment (gitcliff) [reply src/transformers/default-schema-transformer.ts:149]: > What happens if the configured format is not a concept? Maybe a "org.openmrs.util.AttributableDate"?

thanks @samuelmale 
my assumption on this is If the configured format is not `org.openmrs.Location` , the function will still apply the default datasource configuration `(select_concept_answers_datasource)`

Comment (samuelmale) [reply src/transformers/default-schema-transformer.ts:149]: Don't we want to log some kind of error message?

Comment (gitcliff) [reply src/transformers/default-schema-transformer.ts]: Thanks ,,,i added an error message

File: src/api/api.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import { fhirBaseUrl, openmrsFetch, restBaseUrl } from '@openmrs/esm-framework';
 import { encounterRepresentation } from '../constants';
-import { type OpenmrsForm, type PatientIdentifier, type PatientProgramPayload } from '../types';
+import { type PersonAttribute, type OpenmrsForm, type PatientIdentifier, type PatientProgramPayload } from '../types';
```
Comment (NethmiRodrigo) [inline src/api/api.ts:3]:
```suggestion
import type { PersonAttribute, OpenmrsForm, PatientIdentifier, PatientProgramPayload } from '../types';
```

File: src/transformers/default-schema-transformer.ts [hunk 5]
```diff
@@ -35,10 +36,10 @@ function handleQuestion(question: FormField, form: FormSchema) {
   try {
     sanitizeQuestion(question);
     setFieldValidators(question);
-    transformByType(question);
+     transformByType(question);
```
Comment (NethmiRodrigo) [inline src/transformers/default-schema-transformer.ts:39]: Shouldn't this be called with `await`, now that we've made it an async function?

File: src/submission-handlers/personAttributeHandler.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    return null;
+  },
+
+  getDisplayValue: (field: FormField, value: any) => {
+    return value?.display;
+  },
+  getPreviousValue: (field: FormField, encounter: OpenmrsEncounter, allFormFields: Array<FormField>) => {
+    return null;
+  },
```
Comment (NethmiRodrigo) [inline src/submission-handlers/personAttributeHandler.ts:50]: Maybe we could just do 
```suggestion
  getPreviousValue: () => null,
```

**Comment by brandones** [comment]
@gitcliff Are you addressing @NethmiRodrigo 's feedback?

**Comment by brandones** [comment]
@gitcliff Please fix the conflicts. And please post a comment when this is ready for review, or click the "Re-request Review" button next to Nethmi's name if that button is available to you:

image

**Comment by gitcliff** [comment]
> @gitcliff Please fix the conflicts. And please post a comment when this is ready for review, or click the "Re-request Review" button next to Nethmi's name if that button is available to you:
> 
> image

yes sure ,,, thanks @brandones

**Comment by brandones** [comment]
Ping @gitcliff , what's the status?

**Comment by samuelmale** [comment]
Closing this for now. @gitcliff feel free to resume working on this when you get the suspense.

**Pull request closed without merging**